### PR TITLE
Reduce unneeded github workflow runs

### PR DESCRIPTION
**Pull request by roy-dydx**
### Changelist
Removed some unneeded workflow runs.
- Indexer/protocol test workflows run when there are changes under the correct directory and on appropriate release branches
- All ECR builds still happen for release branches. This is still needed due to how our deploy works at the moment. Consider improving in the future.

### Test Plan
N/A

### Author/Reviewer Checklist
- [ ] If this change affects functionality (features, bug fixes, breaking changes, etc.), update `protocol/CHANGELOG.md` and/or `indexer/CHANGELOG.md` appropriately.
- [ ] If this PR has changes that result in a different app state given the same prior state and transaction list, manually add the `state-breaking` label.
- [ ] If this PR isn't state-breaking but has changes that modify behavior in `PrepareProposal` or `ProcessProposal`, manually add the label `proposal-breaking`.
- [ ] If this PR is one of many that implement a specific feature, manually label them all `feature:[feature-name]`.
- [ ] If you wish to for mergify-bot to automatically create a PR to backport your change to a release branch, manually add the label `backport/[branch-name]`.
- [x] Manually add any of the following labels: `refactor`, `chore`, `bug`.

### Summary by CodeRabbit

- Chore: Updated GitHub Actions workflows to refine triggering conditions based on specific release branches and directories. This includes changes to branch filtering patterns and path filters in multiple workflow files, allowing for more precise control over when workflows are triggered.
- New Feature: Added new event triggers for pull requests on the `protocol` and `indexer` directories, ensuring relevant workflows run when changes are made in these areas.
- Refactor: Updated concurrency groups in several workflows to include the workflow name and either the head reference or the run ID, improving workflow management and preventing unnecessary runs.

### CODE REVIEW

File: .github/workflows/protocol-sim.yml [hunk 1]
```diff
@@ -2,11 +2,14 @@ name: Protocol Sims
...
 on:  # yamllint disable-line rule:truthy
+  pull_request:
+    paths:
+      - 'protocol/**'
   push:
     branches:
       - main
-      - 'release/[a-z]+/v[0-9]+.[0-9]+.x'  # e.g. release/ibctestnet/v1.0.x
-      - 'release/v[0-9]+.[0-9]+.x'  # e.g. release/v0.0.x
+      - 'release/protocol/v0.[0-9]+.x'  # e.g. release/protocol/v0.1.x
```
Comment (teddyding) [inline .github/workflows/protocol-sim.yml:11]: This doesn't prevent the scenario described [here](https://dydx-team.slack.com/archives/GTGEVSCUU/p1696523249079489) right? 

> Protocol Sims are run on indexer only commits ([e.g.](https://github.com/dydxprotocol/v4-chain/actions/runs/6357582054)), which takes ~40min

Can we do something like 


```
  push:
    paths: 
      - 'protocol/**'
    branches:
      - main
      - 'release/protocol/v0.[0-9]+.x'  # e.g. release/protocol/v0.1.x
```

Is it an "or" or "and" logic between `paths` and `branches` here?

Comment (roy-dydx) [reply .github/workflows/protocol-sim.yml:11]: Ah oops, nice catch.

File: .github/workflows/indexer-build-and-push-testnet.yml [hunk 2]
```diff
@@ -4,8 +4,8 @@ on: # yamllint disable-line rule:truthy
   push:
     branches:
       - releases/testnet
-      - 'release/[a-z]+/v[0-9]+.[0-9]+.x'  # e.g. release/ibctestnet/v1.0.x
-      - 'release/v[0-9]+.[0-9]+.x'  # e.g. release/v0.0.x
+      - 'release/[a-z]+/0.[0-9]+.x'  # e.g. release/indexer/v0.1.x
+      - 'release/[a-z]+/v[0-9]+.x'  # e.g. release/indexer/v1.x
```
Comment (teddyding) [inline .github/workflows/indexer-build-and-push-testnet.yml:8]: To confirm, we explicitly still want build and push workflow on every push to `main` right?

File: .github/workflows/indexer-api-documentation-check.yml [hunk 3]
```diff
@@ -6,8 +6,10 @@ on:  # yamllint disable-line rule:truthy
   push:
     branches:
       - main
-      - 'release/[a-z]+/v[0-9]+.[0-9]+.x'  # e.g. release/ibctestnet/v1.0.x
-      - 'release/v[0-9]+.[0-9]+.x'  # e.g. release/v0.0.x
+      - 'release/indexer/v0.[0-9]+.x'  # e.g. release/indexer/v0.1.x
+      - 'release/indexer/v[0-9]+.x'  # e.g. release/indexer/v1.x
```
[the remark this answers is not in the document]
Comment (roy-dydx) [reply .github/workflows/indexer-api-documentation-check.yml:10]: this ain't a regex

File: .github/workflows/indexer-build-and-push-testnet.yml (hunk 2 above)
Comment (roy-dydx) [reply .github/workflows/indexer-build-and-push-testnet.yml:8]: Yeah. I don't know if we really use this environment anymore tbh. I'll just put main for now.

File: .github/workflows/indexer-api-documentation-check.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (roy-dydx) [reply .github/workflows/indexer-api-documentation-check.yml:10]: I'm not sure about that. Period isn't a special character here.

**Review by teddyding — APPROVED** [review]

**Pull request merged**